keps/sig-api-machinery/20190820-master-service-of-type-externalname.md

```diff
@@ -0,0 +1,304 @@
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+The master kubernetes service created in the `default` namespace is called `kubernetes`. It's used by Pods running in the cluster to discover the kube-apiserver and communicate with it.
+
+kube-apiserver manages the master `Service` of type `ClusterIP` and associated `Endpoints` objects. Using this service, kubelet injects two environment variables called `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT` into Pod's containers. Those environment variables can be detected by Kubernetes API clients such as `client-go` and used to determine if they run as a Pod in the cluster.
+
+This KEP is about adding support in Kubernetes for the master service to be of type `ExternalName` in addition to `ClusterIP`. This allows for kube-apiservers exposed via external FQDN (fully qualified domain name) to be directly consumed by Pods running in the cluster via that FQDN.
+
+## Motivation
+
+In a Kubernetes cluster, the kube-apiserver(s) can be reached via its IPv4 and/or IPv6 address (e.g. `10.0.0.1`) and optionally by its hostname (e.g. `master.local`) from out-of-cluster clients. The apiserver(s) is started with `--advertise-address=10.0.0.1`. This IP address is used for in-cluster apiserver discovery.
+
+If this IP address is changed, a reconfiguration and restart of the apiserver(s) is required. Until that operation is performed, in-cluster workload using automatic apiserver discovery would send traffic to the old and invalid `10.0.0.1` address.
```

Comment:
This is the sticking point for me.

The "advertise address" is supposed to fill in an Endpoints, but clients should never see that. They see the ClusterIP, which is a virtualization of the set of advertise addresses of all masters.

The only components which CAN'T use this mechanism are:
* kubelet (it starts everything)
* kube-proxy (it implements the VIP abstraction)
* anything running in hostNetwork

Both kubelet and kube-proxy can or could be configured to use hostnames.

So are we trying to solve for hostNetwork?

Reply:
> So are we trying to solve for hostNetwork?

I was mainly solving this for cluster administrators which are forced to react on such changes and optionally perform manual operations.

Reply:
My point was that we can define the kubelet and kube-proxy configs to take hostnames.  They would use this named-LB.  Other pods don't need cluster admin help - they use the service VIP.

I grant that this does not allow for the master service to be defined in terms of a name-oriented LB for pods.  I consider that a shortcoming, but very hard to mitigate with the legacy of KUBERNETES_SERVICE_HOST having always been an IP.